### PR TITLE
 Python: indicate whether a derivation provides a Python module

### PR DESCRIPTION
**Pull request by FRidh**
###### Motivation for this change

The `python-packages.nix` file contains derivations that provide Python modules. Sometimes, one needs to determine whether packages provide Python modules. Currently, a hook is used for this, but the hook does not check whether the `site-packages` correspond to the right Python derivation.

Derivations providing Python modules should be marked with `isPythonModule = python;` where `python` is the interpreter that was used to build the module.

Most of the derivations in `python-packages.nix` are produced with `buildPythonPackage` but not all of them. Therefore, a helper function `toPythonModule` is provided that sets this attribute, and corrects the name as well.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by grahamc** [comment]
:)

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -80,10 +84,21 @@ let
         else throw "Unsupported kind ${kind}");
     in fetcher (builtins.removeAttrs attrs ["format"]) );
 
+  # Check whether a derivation provides a Python module.
+  hasPythonModule = drv: (hasAttr "pythonModule" drv) && ( (getAttr "pythonModule" drv) == python);
+
+  # Recurse into a list of Python modules, returning all Python modules that are required.
+  requiredModules = drvs: filter hasPythonModule (closePropagation drvs);
```
Comment (veprbl) [inline pkgs/top-level/python-packages.nix]: Maybe name it ```requiredPythonModules``` or even ```filterPythonModules```. The meaning of ```requiredModules``` is hard to understand if you don't think in the context of Python modules.

**Comment by veprbl** [comment]
When I first tried python package using ```mkDerivation``` I was really surprised that
```nix
stdenv.mkDerivation {
  #...
  buildInputs = with pythonPackages; [ foo ];
}
```
is not automatically transformed into
```nix
stdenv.mkDerivation rec {
  #...
  buildInputs = with pythonPackages; [ foo ];
  pythonPath = requiredModules buildInputs;
}
```
Or perhaps even do this in bash. Say, we could call ```buildPythonPath()``` on ```$buildInputs``` instead of ```$pythonPath```. I guess there must be a good reason to not do that. Is it because it can be too slow?

**Comment by FRidh** [comment]
@veprbl certain inputs are only needed at build or test time. Also, non-python dependencies should not be added to `PYTHONPATH`.

Certain things can be done in `sh` but I'm actually trying to get rid of that because its brittle and because you cannot guarantee that your module belongs to the interpreter for which you seek the modules.

**Comment by veprbl** [comment]
@FRidh Things that are only needed at build and test time go to ```nativeBuildInputs```. My point was that it would be nice if things would work automagically with what is put in ```buildInputs``` like it already does for many different places in nixpkgs. As far as I understand, it usually is implemented via shell hooks, which is why I brought up shell. It can be also implemented fully in nix, like you did here, except that nix doesn't have information whether a dependence will produce "${drv}/lib/pythonXX/" path or not, so you need to introduce ```toPythonModule``` to explicitly mark each and every derivation manually.

Also, ```yoda``` and ```rivet``` could use ```toPythonModule``` too.

**Comment by teto** [comment]
I had to clean up some space and disable some packages that were too long to recompile but finally ended up ok. It fixed the problem for me (1st version of the PR, you 've pushed a new version since then but I can't recompile on every push sry xD).
`Also, non-python dependencies should not be added to PYTHONPATH.`
couldn't buildInputs identify python packages via checking for isPython in the atttSet ? 
Not sure what to do but  ditching (i.e., automate the generation) pythonPath for buildInputs would make packaging easier.

**Comment by FRidh** [comment]
>  It fixed the problem for me

Good to hear. 

> couldn't buildInputs identify python packages via checking for isPython in the atttSet ?

Yes and no. Currently the Python interpreter comes with a setup hook which adds the site-packages folders to `PYTHONPATH`. The issue is that it does not distinguish between interpreter versions.

I would prefer to do most of this in Nix instead of through a  setup hook. In case of our `buildPythonPackage` function and `python.buildEnv` we can do that easily. But in case of `stdenv.mkDerivation` we cannot.

**Comment by FRidh** [comment]
Open issues: 
- `python.pkgs.libxml2` used to give the `.py` output of `libxml2`. Application of `toPythonModule` changed that to `bin`.

**Comment by FRidh** [comment]
I've pushed this to staging.

**Pull request closed without merging**